simulator/bypass/data_bypass.h

```diff
@@ -0,0 +1,172 @@
+/**
+ * data_bypass.h - Implementation of forwarding unit
+ * 
+ * @author Denis Los
+ * Copyright 2018 MIPT-MIPS Project
+ */
+
+
+#ifndef DATA_BYPASS_H
+#define DATA_BYPASS_H
+
+
+#include <array>
+
+#include <mips/mips_instr.h>
+#include <infra/ports/timing.h>
+
+
+
+class DataBypass
+{
+    public:
+        class RegisterStage
+        {
+            public:
+                explicit RegisterStage( uint8 value) : value( value) { }
+
+                auto operator==( const RegisterStage& rhs) const { return value == rhs.value; }
+                explicit operator std::size_t() const { return static_cast<std::size_t>( value); }
+
+                void inc() { ++value; }
+
+                constexpr auto get_latency_from_first_execute_stage() const { return Latency( value); }
+
+                static constexpr auto get_bypassing_stages_number() { return BYPASSING_STAGES_NUM; }
+                static RegisterStage in_RF() { return RegisterStage( IN_RF_STAGE_VALUE); }
+
+            private:
+                uint8 value = 0;  // distance from first execute stage
+
+                // EXECUTE   - 0  | Bypassing stage
+                // MEMORY    - 1  | Bypassing stage
+                // WRITEBACK - 2  | Bypassing stage
+                // IN_RF     - 3
+
+                static constexpr const uint8 IN_RF_STAGE_VALUE = 3;
```

Comment:
I would use MAX_UINT8 so there would be no need to increment it if more stages are added.